### PR TITLE
arch: riscv: stacktrace support with stack pointer

### PR DESCRIPTION
**Pull request by ycsin**
Add sp-based stack unwinding implementation, when compiling the test application, the size is reduced by ~4kB when compared to the fp-based implementation

### CODE REVIEW

**Comment by npitre** [comment]
One concern is about `get_sp()` which is way too simple a symbol for a
global scope. You should consider `z_riscv_get_sp_before_exc()` or
the like.

**Review by cfriedt — DISMISSED** [review]

**Comment by ycsin** [comment]
> z_riscv_get_sp_before_exc

Updated `get_sp` to `z_riscv_get_sp_before_exc` and `unwind_stack` to `z_riscv_unwind_stack`

**Review by fkokosinski — CHANGES_REQUESTED** [review]
Small nit, the rest LGTM

File: arch/riscv/core/fatal.c [hunk 1]
```diff
@@ -28,82 +28,29 @@ static const struct z_exc_handle exceptions[] = {
...
-		if (in_text_region(ra)) {
-			LOG_ERR("     %2d: fp: " PR_REG "   ra: " PR_REG, i, fp, ra);
-			/*
-			 * Increment the iterator only if `ra` is within the text region to get the
-			 * most out of it
-			 */
-			i++;
-		}
-		fp = frame->fp;
+	if ((esf->mstatus & MSTATUS_MPP) == 0) {
```
[the remark this answers is not in the document]
Comment (ycsin) [reply arch/riscv/core/fatal.c]: fixed

**Review by fkokosinski — APPROVED** [review]

File: arch/riscv/core/stacktrace.c [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+LOG_MODULE_DECLARE(os, CONFIG_KERNEL_LOG_LEVEL);
+
+uintptr_t z_riscv_get_sp_before_exc(const z_arch_esf_t *esf);
+
+#if __riscv_xlen == 32
+ #define PR_REG "%08" PRIxPTR
+#elif __riscv_xlen == 64
+ #define PR_REG "%016" PRIxPTR
+#endif
```
Comment (cfriedt) [inline arch/riscv/core/stacktrace.c:20]: I would guess that `%p` would prefix with `0x` and and zero-pad values appropriately, but then they would need to be cast to `(void *)` as well.

Maybe the following would be better? It skips the `0x` prefix though (but at least there is no cast required. 
```suggestion
#if __riscv_xlen == 32
 #define PR_REG "%08" PRIx32
#elif __riscv_xlen == 64
 #define PR_REG "%016" PRIx64
#endif
```

**Review by cfriedt — APPROVED** [review]
LGTM - nothing blocking

**Review by npitre — APPROVED** [review]

File: arch/riscv/core/stacktrace.c (hunk 2 above)
Comment (ycsin) [reply arch/riscv/core/stacktrace.c:20]: Will take a look there when I do the symtab PR after this

Comment (cfriedt) [reply arch/riscv/core/stacktrace.c:20]: Sure - this isn't really blocking though, so if it works, it works.

Comment (ycsin) [reply arch/riscv/core/stacktrace.c:20]: The current implementation works without casting, `PRIx64`/`PRIx32` have to be casted

**Pull request merged**